src/lib/FreeC/Pass/EtaConversionPass.hs

```diff
@@ -106,57 +132,98 @@ etaConversionPass ast = do
 -- Function declarations                                                     --
 -------------------------------------------------------------------------------
 
--- | Applies 'etaConvertTopLevelRhs' to the right-hand side of the given function
---   declaration to ensure all functions and constructors on the right-hand side 
+-- | Depending on the presence or absence of missing top-level arguments, applies 
+--   'modifyTopLevel' or 'etaConvertExpr' to the right-hand side of the given 
+--   function declaration to ensure all functions and constructors on the right-hand side 
 --   are fully applied. 
 --   The missing top-level arguments are also added to the left-hand side of the 
 --   declaration and the function's type and the environment are updated accordingly. 
 etaConvertFuncDecl :: IR.FuncDecl -> Converter IR.FuncDecl
 etaConvertFuncDecl funcDecl = do
   let rhs = IR.funcDeclRhs funcDecl
-  newArgIdents <- generateFullArgumentList rhs
-  -- Compute the function's new (uncurried) type. Assumes that funcDecl's return type has already been inferred.
-  let (newArgTypes, returnType) = IR.splitFuncType
-        (fromJust $ IR.funcDeclReturnType funcDecl)
-        (length newArgIdents)
+  newArgNumber <- findMinMissingArguments rhs
```

Comment:
I'm currently regretting reading the file bottom to top instead of the other way around : ) I now have the impression that `findMinMissingArguments` and `arityOf` do not need to be monadic (i.e., of type `Converter <something>`) -- there's not context needed to compute the needed value. Can you confirm or refute that impression?

Reply:
`arityOf` needs to look up the arity of a function in the current environment, so it does need to produce a value inside the `Converter` monad. 
`inEnv` has the type `MonadConverter m => (Environment -> a) -> m a`. 
Since `findMinMissingArguments` uses `arityOf`, it also needs to produce a monadic value.

Reply:
Ah, I see. I missed that from the latest diff I was looking at!